state/spaces.go

```diff
@@ -81,25 +91,82 @@ func (s *Space) Subnets() (results []*Subnet, err error) {
 }
 
 // AddSpace creates and returns a new space.
-func (st *State) AddSpace(name string, providerId network.Id, subnets []string, isPublic bool) (newSpace *Space, err error) {
+func (st *State) AddSpace(
+	name string, providerId network.Id, subnets []string, isPublic bool) (newSpace *Space, err error,
+) {
 	defer errors.DeferredAnnotatef(&err, "adding space %q", name)
 	if !names.IsValidSpace(name) {
 		return nil, errors.NewNotValid(nil, "invalid space name")
 	}
 
-	spaceDoc := spaceDoc{
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if _, err := st.Space(name); err != nil {
+			if !errors.IsNotFound(err) {
+				return nil, errors.Annotatef(err, "checking for existing space")
+			}
+		} else {
+			return nil, errors.AlreadyExistsf("space %q", name)
+		}
+
+		for _, subnetId := range subnets {
+			subnet, err := st.Subnet(subnetId)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if subnet.FanLocalUnderlay() != "" {
+				return nil, errors.Errorf(
+					"cannot set space for FAN subnet %q - it is always inherited from underlay", subnet.CIDR())
+			}
+		}
+
+		// The ops will assert that the ID is unique,
+		// but we check explicitly in order to return an indicative error.
+		if providerId != "" {
+			exists, err := st.networkEntityGlobalKeyExists("space", providerId)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if exists {
+				return nil, errors.Errorf("provider ID %q not unique", providerId)
+			}
+		}
+
+		ops, err := st.addSpaceTxnOps(name, providerId, subnets, isPublic)
+		return ops, errors.Trace(err)
+	}
+
+	err = st.db().Run(buildTxn)
+	if err != nil {
+		err = onAbort(err, ErrDead)
+		logger.Errorf("cannot add space to the model: %v", err)
+		return nil, errors.Trace(err)
+	}
+
+	space, err := st.Space(name)
+	return space, errors.Trace(err)
+}
+
+func (st *State) addSpaceTxnOps(name string, providerId network.Id, subnets []string, isPublic bool) ([]txn.Op, error) {
+	// Space with ID zero is the default space; start at 1.
+	seq, err := sequenceWithMin(st, "space", 1)
+	if err != nil {
+		return nil, err
+	}
+	id := strconv.Itoa(seq)
+
+	doc := spaceDoc{
+		Id:         id,
 		Life:       Alive,
 		Name:       name,
 		IsPublic:   isPublic,
 		ProviderId: string(providerId),
 	}
-	newSpace = &Space{doc: spaceDoc, st: st}
 
 	ops := []txn.Op{{
 		C:      spacesC,
-		Id:     name,
+		Id:     id,
```

Comment:
If we don't specify DocId at all, I believe you end up getting a bson.ObjectId, which means the data ends up effectively randomly sorted in the database. (I think its actually time based.)

Reply:
I do realize that documents like Machine have both DocID and "machineid", where the former is <modeluuid:id> in the database, while the latter is just "id". However, in-memory it is also just <id>.
I can live with following that pattern (it is probably more predictable). But if we do so, then I would add:
spaceDoc{
...
DocId: st.docID(id),
}
And I would recommend using it in the txn.Op:

Id:     mdoc.DocID,

Note that "Id: id" does "work" because multiModelRunner will translate if necessary.

Reply:
I've made this explicit as suggested.